File: lambda/typescript/src/users-handler.ts
/**
 * Users Lambda Handler - TypeScript Implementation
 * Author: Ali Ucer
 * Created: 2024
 * Description: REST API endpoints for user management with idempotency and retry logic
 */

import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import { DynamoDBDocumentClient, GetCommand, PutCommand, UpdateCommand, DeleteCommand } from '@aws-sdk/lib-dynamodb';
import { APIGatewayProxyEvent, APIGatewayProxyResult, Context } from 'aws-lambda';

// Initialize clients
const client = new DynamoDBClient({});
const docClient = DynamoDBDocumentClient.from(client);
const tableName = process.env.USERS_TABLE_NAME!;

interface User {
  userId: string;
  email: string;
  name?: string;
  department?: string;
  createdAt: number;
  ttl: number;
  updatedAt?: number;
}

interface Response {
  statusCode: number;
  body: string;
  headers?: Record<string, string>;
}

async function sleep(ms: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, ms));
}

async function handler(event: APIGatewayProxyEvent, context: Context): Promise<APIGatewayProxyResult> {
  const method = event.httpMethod;
  const pathParameters = event.pathParameters || {};
  const userId = pathParameters.id;
  
  console.log('Request received', { path: event.path, method, userId });
  
  const body = event.body ? JSON.parse(event.body) : {};
  
  try {
    switch (method) {
      case 'GET':
        if (userId) {
          return await getUser(userId);
        }
        break;
      case 'POST':
        return await createUser(body);
      case 'PUT':
        if (userId) {
          return await updateUser(userId, body);
        }
        break;
      case 'DELETE':
        if (userId) {
          return await deleteUser(userId);
        }
        break;
    }
    
    return {
      statusCode: 405,
      body: JSON.stringify({ error: 'Method not allowed' })
    };
  } catch (error: any) {
    console.error('Handler error', error);
    return {
      statusCode: 500,
      body: JSON.stringify({ error: 'Internal server error' })
    };
  }
}

async function getUser(userId: string): Promise<APIGatewayProxyResult> {
  console.log('Getting user', { userId });
  
  const retries = 3;
  let backoff = 100;
  
  for (let attempt = 0; attempt < retries; attempt++) {
    try {
      const command = new GetCommand({
        TableName: tableName,
        Key: { userId }
      });
      
      const response = await docClient.send(command);
      
      if (!response.Item) {
        return {
          statusCode: 404,
          body: JSON.stringify({ error: 'User not found' })
        };
      }
      
      return {
        statusCode: 200,
        body: JSON.stringify(response.Item)
      };
    } catch (error) {
      if (attempt < retries - 1) {
        await sleep(backoff * Math.pow(2, attempt));
        console.warn('Retrying getUser', { attempt: attempt + 1 });
        continue;
      }
      throw error;
    }
  }
  
  return {
    statusCode: 500,
    body: JSON.stringify({ error: 'Failed to get user' })
  };
}

async function createUser(body: any): Promise<APIGatewayProxyResult> {
  const userId = body.userId;
  const email = body.email;
  
  if (!userId || !email) {
    return {
      statusCode: 400,
      body: JSON.stringify({ error: 'userId and email are required' })
    };
  }
  
  console.log('Creating user', { userId, email });
  
  try {
    // Check if user already exists (idempotency)
    const getCommand = new GetCommand({
      TableName: tableName,
      Key: { userId }
    });
    
    const existing = await docClient.send(getCommand);
    
    if (existing.Item) {
      console.log('User already exists', { userId });
      return {
        statusCode: 200,
        body: JSON.stringify({ message: 'User already exists', user: existing.Item })
      };
    }
    
    const now = Math.floor(Date.now() / 1000);
    const item: User = {
      userId,
      email,
      createdAt: now,
      ttl: now + (90 * 24 * 60 * 60) // TTL: 90 days
    };
    
    if (body.name) item.name = body.name;
    if (body.department) item.department = body.department;
    
    const command = new PutCommand({
      TableName: tableName,
      Item: item,
      ConditionExpression: 'attribute_not_exists(userId)' // Conditional write for idempotency
    });
    
    await docClient.send(command);
    
    console.log('User created successfully', { userId });
    
    return {
      statusCode: 201,
      body: JSON.stringify({ message: 'User created', user: item })
    };
  } catch (error: any) {
    if (error.name === 'ConditionalCheckFailedException') {
      console.log('User already exists (race condition)', { userId });
      const getCommand = new GetCommand({
        TableName: tableName,
        Key: { userId }
      });
      const existing = await docClient.send(getCommand);
      if (existing.Item) {
        return {
          statusCode: 200,
          body: JSON.stringify({ message: 'User already exists', user: existing.Item })
        };
      }
    }
    throw error;
  }
}

async function updateUser(userId: string, body: any): Promise<APIGatewayProxyResult> {
  console.log('Updating user', { userId });
  
  try {
    const getCommand = new GetCommand({
      TableName: tableName,
      Key: { userId }
    });
    
    const existing = await docClient.send(getCommand);
    
    if (!existing.Item) {
      return {
        statusCode: 404,
        body: JSON.stringify({ error: 'User not found' })
      };
    }
    
    const updateExpr: string[] = ['SET updatedAt = :updatedAt'];
    const exprAttr: any = { ':updatedAt': Math.floor(Date.now() / 1000) };
    
    for (const [key, value] of Object.entries(body)) {
      if (key !== 'userId') {
        updateExpr.push(`${key} = :${key}`);
        exprAttr[`:${key}`] = value;
      }
    }
    
    const command = new UpdateCommand({
      TableName: tableName,
      Key: { userId },
      UpdateExpression: updateExpr.join(', '),
      ExpressionAttributeValues: exprAttr,
      ReturnValues: 'ALL_NEW'
    });
    
    const response = await docClient.send(command);
    
    return {
      statusCode: 200,
      body: JSON.stringify(response.Attributes)
    };
  } catch (error) {
    console.error('DynamoDB update error', error);
    throw error;
  }
}

async function deleteUser(userId: string): Promise<APIGatewayProxyResult> {
  console.log('Deleting user', { userId });
  
  try {
    const command = new DeleteCommand({
      TableName: tableName,
      Key: { userId }
    });
    
    await docClient.send(command);
    
    return {
      statusCode: 200,
      body: JSON.stringify({ message: 'User deleted successfully' })
    };
  } catch (error) {
    console.error('DynamoDB delete error', error);
    throw error;
  }
}

export { handler };

File: lambda/python/users_handler.py
"""
Users Lambda Handler
Author: Ali Ucer
Created: 2024
Description: REST API endpoints for user management with idempotency and retry logic
"""

import json
import os
import time
import boto3
import structlog
from typing import Dict, Any
from botocore.exceptions import ClientError

# Initialize clients
dynamodb = boto3.resource('dynamodb')
table = dynamodb.Table(os.environ['USERS_TABLE_NAME'])
logger = structlog.get_logger()

def handler(event: Dict[str, Any], context: Any) -> Dict[str, Any]:
    """
    Lambda handler for users API endpoints
    
    Supports:
    - GET /users/{id} - Get user by ID
    - POST /users - Create user
    - PUT /users/{id} - Update user
    - DELETE /users/{id} - Delete user
    """
    try:
        http_method = event.get('httpMethod')
        request_path = event.get('path', '')
        logger.info("Request received", path=request_path, method=http_method)
        
        method = http_method
        path = request_path
        path_params = event.get('pathParameters') or {}
        body = json.loads(event.get('body', '{}'))
        
        # Extract user ID from path
        user_id = path_params.get('id')
        
        if method == 'GET' and user_id:
            return get_user(user_id)
        elif method == 'POST':
            return create_user(body)
        elif method == 'PUT' and user_id:
            return update_user(user_id, body)
        elif method == 'DELETE' and user_id:
            return delete_user(user_id)
        else:
            return {
                'statusCode': 405,
                'body': json.dumps({'error': 'Method not allowed'})
            }
            
    except json.JSONDecodeError:
        logger.error("Invalid JSON in request body")
        return {
            'statusCode': 400,
            'body': json.dumps({'error': 'Invalid JSON'})
        }
    except ClientError as e:
        logger.error("DynamoDB error", error=str(e))
        return {
            'statusCode': 500,
            'body': json.dumps({'error': 'Internal server error'})
        }
    except Exception as e:
        logger.error("Unexpected error", error=str(e))
        return {
            'statusCode': 500,
            'body': json.dumps({'error': 'Internal server error'})
        }


def get_user(user_id: str) -> Dict[str, Any]:
    """Get user by ID with retry logic"""
    logger.info("Getting user", user_id=user_id)
    
    retries = 3
    backoff = 0.1
    
    for attempt in range(retries):
        try:
            response = table.get_item(
                Key={'userId': user_id}
            )
            
            if 'Item' not in response:
                return {
                    'statusCode': 404,
                    'body': json.dumps({'error': 'User not found'})
                }
            
            return {
                'statusCode': 200,
                'body': json.dumps(response['Item'])
            }
        except ClientError as e:
            if attempt < retries - 1:
                time.sleep(backoff * (2 ** attempt))
                logger.warning("Retrying get_user", attempt=attempt+1)
                continue
            raise
    
    return {'statusCode': 500, 'body': json.dumps({'error': 'Failed to get user'})}


def create_user(body: Dict[str, Any]) -> Dict[str, Any]:
    """Create user with idempotency and conditional write"""
    user_id = body.get('userId')
    email = body.get('email')
    
    if not user_id or not email:
        return {
            'statusCode': 400,
            'body': json.dumps({'error': 'userId and email are required'})
        }
    
    logger.info("Creating user", user_id=user_id, email=email)
    
    try:
        # Check if user already exists (idempotency)
        existing = table.get_item(Key={'userId': user_id})
        if 'Item' in existing:
            logger.info("User already exists", user_id=user_id)
            return {
                'statusCode': 200,
                'body': json.dumps({'message': 'User already exists', 'user': existing['Item']})
            }
        
        # Create user with conditional write to prevent race conditions
        item = {
            'userId': user_id,
            'email': email,
            'createdAt': int(time.time()),
            'ttl': int(time.time()) + (90 * 24 * 60 * 60)  # TTL: 90 days
        }
        
        # Add optional fields
        if 'name' in body:
            item['name'] = body['name']
        if 'department' in body:
            item['department'] = body['department']
        
        response = table.put_item(
            Item=item,
            ConditionExpression='attribute_not_exists(userId)'  # Conditional write for idempotency
        )
        
        logger.info("User created successfully", user_id=user_id)
        
        return {
            'statusCode': 201,
            'body': json.dumps({'message': 'User created', 'user': item})
        }
        
    except ClientError as e:
        if e.response['Error']['Code'] == 'ConditionalCheckFailedException':
            logger.info("User already exists (race condition)", user_id=user_id)
            # Fetch the existing user
            existing = table.get_item(Key={'userId': user_id})
            if 'Item' in existing:
                return {
                    'statusCode': 200,
                    'body': json.dumps({'message': 'User already exists', 'user': existing['Item']})
                }
        raise


def update_user(user_id: str, body: Dict[str, Any]) -> Dict[str, Any]:
    """Update user with optimistic locking"""
    logger.info("Updating user", user_id=user_id)
    
    try:
        # Get existing user
        existing = table.get_item(Key={'userId': user_id})
        
        if 'Item' not in existing:
            return {
                'statusCode': 404,
                'body': json.dumps({'error': 'User not found'})
            }
        
        # Build update expression
        update_expr = "SET updatedAt = :updatedAt"
        expr_attr = {':updatedAt': int(time.time())}
        
        for key, value in body.items():
            if key != 'userId':  # Don't allow changing userId
                update_expr += f", {key} = :{key}"
                expr_attr[f':{key}'] = value
        
        response = table.update_item(
            Key={'userId': user_id},
            UpdateExpression=update_expr,
            ExpressionAttributeValues=expr_attr,
            ReturnValues='ALL_NEW'
        )
        
        return {
            'statusCode': 200,
            'body': json.dumps(response['Attributes'])
        }
        
    except ClientError as e:
        logger.error("DynamoDB update error", error=str(e))
        raise


def delete_user(user_id: str) -> Dict[str, Any]:
    """Delete user"""
    logger.info("Deleting user", user_id=user_id)
    
    try:
        table.delete_item(Key={'userId': user_id})
        
        return {
            'statusCode': 200,
            'body': json.dumps({'message': 'User deleted successfully'})
        }
        
    except ClientError as e:
        logger.error("DynamoDB delete error", error=str(e))
        raise

